scylla/src/transport/topology.rs

```diff
@@ -189,16 +189,20 @@ impl MetadataReader {
 
         // shuffle known_peers to iterate through them in random order later
         self.known_peers.shuffle(&mut thread_rng());
-
-        let address_of_failed_control_connection = self.control_connection_address;
-        let filtered_known_peers = self
-            .known_peers
-            .iter()
-            .filter(|&peer| peer != &address_of_failed_control_connection);
+        debug!(
+            known_peers = self
+                .known_peers
+                .iter()
+                .fold(
+                    String::with_capacity(self.known_peers.len() * 20),
+                    |acc, addr| acc + &addr.to_string()
```

Comment:
You should probably separate the addresses by a whitespace or a comma, otherwise the first octet and the port number will be glued together between consecutive addresses... I guess it is possible to handle adding separators in a `fold`, you will have to account for the empty accumulator case, though.

Alternatively, how about doing `.iter().map(|addr| add.to_string()).collect::<Vec<String>>().join(", ")` ? `collect` and `join` should do two allocations in total - which is more than your code does, but it nicely handles the separator issue.

Ref: https://doc.rust-lang.org/std/primitive.slice.html#method.join